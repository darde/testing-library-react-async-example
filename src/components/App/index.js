import React, { useState, useEffect } from 'react';
import GetPhotos from '../../repositories/unsplash_repository';
import './styles.css';

function App() {
  const [images, setImages] = useState([]);


  useEffect(() => {
    const fetchData = async () => {
      const fetchedImages = await GetPhotos();
      setImages(fetchedImages);
    };
    fetchData();
    console.log('use effect!');
  },[]);

  if (images.length === 0) {
    return <span data-testid="loading">Loading...</span>
  }
  
  return (
    <div className="App">
      <header className="App-header">
        <h1>Images from Unsplash</h1>
      </header>
      <div className="wrapper">
        {
          images.length > 0 && (
            <ul>
              {
                images.map(image => (
                  <li key={image.id}>
                    <img src={image.urls.thumb} alt={image.alt_description} />
                  </li>
                ))
              }
            </ul>
          )
        }
      </div>
    </div>
  );
}

export default App;
