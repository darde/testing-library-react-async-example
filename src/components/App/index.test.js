import React from 'react';
import {render, cleanup, waitForElement, act } from '@testing-library/react';
import mockRepository from '../../repositories/unsplash_repository';
import App from '.';

jest.mock('../../repositories/unsplash_repository');

afterEach(cleanup);

beforeEach(() => {
  mockRepository.mockClear();
});

describe('when the App component is not yet mounted', () => {
  it('renders the loading text', () => {
    const { getByTestId } = render(<App />);

    expect(getByTestId('loading')).toHaveTextContent('Loading...');
  });
})

it('renders title', async () => {
  const { getByText } = render(<App />);

  const titleElement = await waitForElement(() => getByText(/Images from Unsplash/i));
  expect(titleElement).toBeInTheDocument();
});

it('renders a list of images', async () => {
  mockRepository.mockImplementationOnce(() => [
    {
      id: 1,
      alt_description: 'Photo of beach!',
      urls: {
        thumb: 'http://beach.jpg',
      },
    },
    {
      id: 2,
      alt_description: 'Photo of boat!',
      urls: {
        thumb: 'http://boat.jpg',
      },
    },
  ]);

  const { container } = await waitForElement(() => render(<App />));

  expect(container.querySelector('ul')).toBeInTheDocument();
  expect(mockRepository).toHaveBeenCalledTimes(1);
});
