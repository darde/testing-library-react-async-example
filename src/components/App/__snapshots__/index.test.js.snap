// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders unsplash app correctly 1`] = `
<DocumentFragment>
  <span
    data-testid="loading"
  >
    Loading...
  </span>
</DocumentFragment>
`;
